reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,204 @@
+
+The purpose of Ray on spark autoscaling is to support following scenarios:
+
+- For a standalone mode apache spark cluster, assuming multiple users share the spark cluster, and one user submits a spark application that runs a Ray on spark cluster. If Ray on spark autoscaling is supported, and the spark application enables [dynamic resources allocation](https://spark.apache.org/docs/3.4.1/job-scheduling.html#dynamic-resource-allocation), then the Ray on spark cluster starts with zero Ray worker nodes. When Ray jobs are submitted, the Ray on spark cluster requests more Ray worker nodes based on demands (i.e. scales up). It will launch more spark executors for this spark application. After Ray jobs complete, then the Ray on spark cluster scales down. It will terminate idle Ray worker nodes and then trigger some idle spark executor termination. So that with Ray on spark autoscaling feature enabled, when the Ray on spark cluster is idle, it does not occupy spark worker resources.
+
+- For a databricks spark cluster that enables [databricks cluster autoscaling](https://docs.databricks.com/en/clusters/cluster-config-best-practices.html#autoscaling), with Ray on spark autoscaling enabled, we can start the databricks spark cluster with zero initial spark workers. When Ray jobs are submitted, it triggers Ray on spark cluster scaling up and the underlying databricks spark cluster scaling up. When Ray jobs complete, it triggers Ray on spark cluster scaling down and the underlying databricks spark cluster scaling down. This feature helps databricks users to improve cloud resource utilization and save cost.
+
+
+#### Key requirements:
+- Supports Apache Spark cluster that is configured with [standalone mode](https://spark.apache.org/docs/3.4.1/spark-standalone.html).
+- Support Apache Spark application that enables dynamic resource allocation.
+- Supports databricks spark cluster that enables spark worker autoscaling.
+
+
+### Should this change be within `ray` or outside?
+Within Ray. For better code maintenance.
+
+
+## Stewardship
+### Required Reviewers
+
+- @jjyao
+- @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+- @jjyao
+
+## Design and Architecture
+
+### Prerequisites
+
+- The user has an active Spark cluster (apache/spark >= 3.3).
+
+- The user must install Ray packages on every node of the Spark cluster, e.g. via:
+
+  ```
+  pip install ray
+  ```
+
+  If the user's workload requires Ray add-ons, they must also be installed, e.g. via
+
+  ```
+  pip install "ray[debug,dashboard,tune,rllib,serve]"
+  ```
+
+### Ray on spark autoscaling interfaces
+
+Integrate autoscaling feature into existing `ray.util.spark.cluster_init.setup_ray_cluster` API,
+The following new arguments are added:
+
+ - autoscale: bool (default False)
```

Comment:
Do we ever would want to start an autoscaling cluster with some non-zero worker counts?

Reply:
In my test, launching a Ray worker node is very fast, so I think starting with zero worker node should fulfill most use-cases. If user requires min worker node number > 0, we can support it in future.